### DISCUSSION TITLE
Changing username on oAuth signup

### DISCUSSION THREAD

**Discussion by mrpn** [pocketbase/pocketbase#3723] [Q&A] [answered]
Hello, I'm setting up oAuth with Discord. When someone signs up, their Discord `username` is being used for their PB username. Instead of value from `username`, how can I use Discord `id` as a username?  I've tried to change it using `onRecordBeforeAuthWithOAuth2Request` hook, but I couldn't figure out.

log from: `e.oAuth2User`
```js
{
  "id": "12324535939933",
  "name": "userzz#0",
  "username": "userzz",
  "email": "",
  "avatarUrl": "https://cdn.discordapp.com/avatars/12324535939933/.png",
  "rawUser": {
    "premium_type": 0,
    "discriminator": "0",
    "avatar": null,
    "public_flags": 0,
    "flags": 0,
    "banner": null,
    "locale": "en-US",
    "username": "userzz",
    "accent_color": null,
    "global_name": "user man",
    "avatar_decoration_data": null,
    "banner_color": null,
    "mfa_enabled": false,
    "id": "12324535939933"
  },
  "accessToken": "asdasdasd",
  "refreshToken": "asdasdasdasd"
}
```

more context [auth/discord.go](https://github.com/pocketbase/pocketbase/blob/4abe199acc6836545649db80d94a8257c8a44886/tools/auth/discord.go)

**Comment by ganigeorgiev** [comment, marked as answer]
One slightly hacky workaround could be to reassign in the before hook the OAuth2 user's "username" field, something like:

```js
onRecordBeforeAuthWithOAuth2Request((e) => {
    if (e.providerName == "discord") {
        e.oAuth2User.username = e.oAuth2User.id;
    }
})
```

**Reply by mrpn** [reply]
Thanks, the solution partially worked, but now the new username has three random numbers added to the Discord ID. I've checked my users' collection and confirmed that the ID is unique.

Then, suspecting there might be a BigInt issue, I also tried the following method but still random numbers added to the username. 

```js
onRecordBeforeAuthWithOAuth2Request((e) => {
    if (e.providerName == "discord") {
        e.oAuth2User.username = e.oAuth2User.id.toString();
    }
}, "users");
```

**Reply by ganigeorgiev** [reply]
Additional random part should be added only if the there is an existing user with the provided username.

Also note that the oauth2 username will be used as PocketBase record username only during sign-up/create.

If the above doesn't work, an alternative approach could be to attach to the after hook and update the record manually.

```js
onRecordAfterAuthWithOAuth2Request((e) => {
    if (!e.isNewRecord || e.providerName != "discord") {
        return;
    }

    e.record.set("username", e.oAuth2User.id)
    $app.dao().saveRecord(e.record);
}, "users")
```

**Reply by mrpn** [reply]
You are absolutely right, my username was not unique. 

I thought I checked that the uniqueness, simply by popping the Discord ID into the search bar in admin panel. However, when I actually queried `username = '123213'` it turns out that record was already there. 

My bad, sorry 😞
